futures-util/src/task/atomic_waker.rs

```diff
@@ -198,80 +238,91 @@ impl AtomicWaker {
     /// }
     /// ```
     pub fn register(&self, lw: &LocalWaker) {
-        match self.state.compare_and_swap(WAITING, REGISTERING, Acquire) {
-            WAITING => {
-                unsafe {
-                    // Locked acquired, update the waker cell
-                    *self.waker.get() = Some(lw.clone().into_waker());
-
-                    // Release the lock. If the state transitioned to include
-                    // the `WAKING` bit, this means that a wake has been
-                    // called concurrently, so we have to remove the waker and
-                    // wake it.`
-                    //
-                    // Start by assuming that the state is `REGISTERING` as this
-                    // is what we jut set it to.
-                    let res = self.state.compare_exchange(
-                        REGISTERING, WAITING, AcqRel, Acquire);
-
-                    match res {
-                        Ok(_) => {}
-                        Err(actual) => {
-                            // This branch can only be reached if a
-                            // concurrent thread called `wake`. In this
-                            // case, `actual` **must** be `REGISTERING |
-                            // `WAKING`.
-                            debug_assert_eq!(actual, REGISTERING | WAKING);
-
-                            // Take the waker to wake once the atomic operation has
-                            // completed.
-                            let waker = (*self.waker.get()).take().unwrap();
-
-                            // Just swap, because no one could change state while state == `REGISTERING` | `WAKING`.
-                            self.state.swap(WAITING, AcqRel);
-
-                            // The atomic swap was complete, now
-                            // wake the task and return.
-                            waker.wake();
-                        }
+        let state = self.state.load(Acquire);
+        if (state == EMPTY || state == HOLDS)
+            && self.state.compare_and_swap(state, REGISTERING, Acquire) == state
```

Comment:
It seems that something is missing here...
if self.state.compare_and_swap(state, REGISTERING, Acquire) == WAKING ||
   self.state.compare_and_swap(state, REGISTERING, Acquire) ==  WAKING | HOLDS
the lw.wake() will never be execute? because state is EMPTY or HOLDS...

Reply:
But if this happen, it means that thread A Observed thread B's behaviour, and thread A will execute Its own "lw.wake()" in the original logic, after this PR thread A will not execute...so this "lw.wake()" lost.
Sorry, I'm not sure whether this is right...
CC @cramertj

Reply:
OK, I think now I understand what you mean.

I was reading [this comment](https://github.com/rust-lang-nursery/futures-rs/issues/942#issuecomment-379515834) which says basically that register is `acquire` and `wake` is `release`. This is what I tried to preserve.

However, optimization of `register` was not the goal of this patch (and generally `register` is not on hot path), so I can change `register` back to unconditional CAS (with a little loop).